modules/ml/src/svmsgd.cpp

```diff
@@ -0,0 +1,595 @@
+
+    virtual String getDefaultName() const {return "opencv_ml_svmsgd";}
+
+    virtual void setOptimalParameters(int svmsgdType = ASGD, int marginType = SOFT_MARGIN);
+
+    virtual int getSvmsgdType() const;
+
+    virtual void setSvmsgdType(int svmsgdType);
+
+    virtual int getMarginType() const;
+
+    virtual void setMarginType(int marginType);
+
+    CV_IMPL_PROPERTY(float, Lambda, params.lambda)
+    CV_IMPL_PROPERTY(float, Gamma0, params.gamma0)
+    CV_IMPL_PROPERTY(float, C, params.c)
+    CV_IMPL_PROPERTY_S(cv::TermCriteria, TermCriteria, params.termCrit)
+
+private:
+    void updateWeights(InputArray sample, bool isFirstClass, float gamma, Mat &weights);
+
+    std::pair<bool,bool> areClassesEmpty(Mat responses);
+
+    void writeParams( FileStorage &fs ) const;
+
+    void readParams( const FileNode &fn );
+
+    static inline bool isFirstClass(float val) { return val > 0; }
+
+    static void normalizeSamples(Mat &matrix, Mat &average, float &multiplier);
+
+    float calcShift(InputArray _samples, InputArray _responses) const;
+
+    static void makeExtendedTrainSamples(const Mat &trainSamples, Mat &extendedTrainSamples, Mat &average, float &multiplier);
+
+
+
+    // Vector with SVM weights
+    Mat weights_;
+    float shift_;
+
+    // Parameters for learning
+    struct SVMSGDParams
+    {
+        float lambda;                             //regularization
+        float gamma0;                             //learning rate
+        float c;
+        TermCriteria termCrit;
+        SvmsgdType svmsgdType;
+        MarginType marginType;
+    };
+
+    SVMSGDParams params;
+};
+
+Ptr<SVMSGD> SVMSGD::create()
+{
+    return makePtr<SVMSGDImpl>();
+}
+
+std::pair<bool,bool> SVMSGDImpl::areClassesEmpty(Mat responses)
+{
+    CV_Assert(responses.cols == 1 || responses.rows == 1);
+    std::pair<bool,bool> emptyInClasses(true, true);
+    int limit_index = responses.rows;
+
+    for(int index = 0; index < limit_index; index++)
+    {
+        if (isFirstClass(responses.at<float>(index)))
+            emptyInClasses.first = false;
+        else
+            emptyInClasses.second = false;
+
+        if (!emptyInClasses.first && ! emptyInClasses.second)
+            break;
+    }
+
+    return emptyInClasses;
+}
+
+void SVMSGDImpl::normalizeSamples(Mat &samples, Mat &average, float &multiplier)
+{
+    int featuresCount = samples.cols;
+    int samplesCount = samples.rows;
+
+    average = Mat(1, featuresCount, samples.type());
+    for (int featureIndex = 0; featureIndex < featuresCount; featureIndex++)
+    {
+        Scalar scalAverage = mean(samples.col(featureIndex))[0];
+        average.at<float>(featureIndex) = static_cast<float>(scalAverage[0]);
+    }
+
+    for (int sampleIndex = 0; sampleIndex < samplesCount; sampleIndex++)
+    {
+        samples.row(sampleIndex) -= average;
+    }
+
+    double normValue = norm(samples);
+
+    multiplier = static_cast<float>(sqrt(samples.total()) / normValue);
```

Comment:
Shouldn't, because the metric of the original space will change so that the optimal decision in the resulting space won't be equivalent to the optimal decision in the original space.